core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -0,0 +1,2103 @@
+
+            int attempt = maxTaskExecutionAttempts - partition.getRemainingAttempts();
+            ExchangeSinkInstanceHandle exchangeSinkInstanceHandle = exchange.instantiateSink(partition.getExchangeSinkHandle(), attempt);
+            SpoolingOutputBuffers outputBuffers = SpoolingOutputBuffers.createInitial(exchangeSinkInstanceHandle, sinkPartitioningScheme.getPartitionCount());
+            Optional<RemoteTask> task = stage.createTask(
+                    node,
+                    partitionId,
+                    attempt,
+                    sinkPartitioningScheme.getBucketToPartitionMap(),
+                    outputBuffers,
+                    splits,
+                    noMoreSplits,
+                    Optional.of(partition.getMemoryRequirements().getRequiredMemory()));
+            task.ifPresent(remoteTask -> partition.addTask(remoteTask, outputBuffers));
+            return task;
+        }
+
+        public void updateExchangeSinkInstanceHandle(TaskId taskId)
+        {
+            if (getState().isDone()) {
+                return;
+            }
+            StagePartition partition = getStagePartition(taskId.getPartitionId());
+            ExchangeSinkInstanceHandle exchangeSinkInstanceHandle = exchange.updateSinkInstanceHandle(partition.getExchangeSinkHandle(), taskId.getAttemptId());
+            partition.updateExchangeSinkInstanceHandle(taskId, exchangeSinkInstanceHandle);
+        }
+
+        public void taskFinished(TaskId taskId, TaskStatus taskStatus)
+        {
+            if (getState().isDone()) {
+                return;
+            }
+
+            int partitionId = taskId.getPartitionId();
+            StagePartition partition = getStagePartition(partitionId);
+            exchange.sinkFinished(partition.getExchangeSinkHandle(), taskId.getAttemptId());
+            SpoolingOutputStats.Snapshot outputStats = partition.taskFinished(taskId);
+
+            if (!remainingPartitions.remove(partitionId)) {
+                // a different task for the same partition finished before
+                return;
+            }
+
+            updateOutputSize(outputStats);
+
+            // task tescriptor has been created
+            if (partition.getOpenTaskDescriptor().isEmpty()) {
+                taskDescriptorStorage.remove(stage.getStageId(), partitionId);
+            }
+
+            partitionMemoryEstimator.registerPartitionFinished(
+                    queryStateMachine.getSession(),
+                    partition.getMemoryRequirements(),
+                    taskStatus.getPeakMemoryReservation(),
+                    true,
+                    Optional.empty());
+
+            sinkOutputSelectorBuilder.include(exchange.getId(), taskId.getPartitionId(), taskId.getAttemptId());
+
+            if (noMorePartitions && remainingPartitions.isEmpty() && !stage.getState().isDone()) {
+                // TODO: support task splitting
```

Comment:
what exactly is to be changed here?

Reply:
Probably a copy-paste mistake. I don't think task every has to be split if it finished successfully.